File: src/rnet/packets/ZonePowerPacket.js
const DataPacket = require("./DataPacket");

class ZonePowerPacket extends DataPacket {
    requiresHandshake() {
        return true;
    }

    getControllerID() {
        return this.sourceControllerID;
    }

    getZoneID() {
        return this.sourcePath[2];
    }

    getPower() {
        return this.data.readUInt8(0) == 1;
    }
}

ZonePowerPacket.fromPacket = function(dataPacket) {
    if (dataPacket instanceof DataPacket) {
        const zonePowerPacket = new ZonePowerPacket();
        dataPacket.copyToPacket(zonePowerPacket);
        return zonePowerPacket;
    }
    else {
        throw new TypeError("Cannot create ZonePowerPacket with anything other than a DataPacket");
    }
}

module.exports = ZonePowerPacket;

File: src/RNetPi.Core/RNet/ZonePowerPacket.cs
using System;

namespace RNetPi.Core.RNet;

/// <summary>
/// Represents an RNet Zone Power packet
/// </summary>
public class ZonePowerPacket : DataPacket
{
    public ZonePowerPacket()
    {
        MessageType = 0x00;
    }

    public byte GetControllerID()
    {
        return SourceControllerID;
    }

    public byte GetZoneID()
    {
        return SourcePath.Length > 2 ? SourcePath[2] : (byte)0;
    }

    public bool GetPower()
    {
        return Data.Length > 0 && Data[0] == 1;
    }

    /// <summary>
    /// Creates a ZonePowerPacket from a DataPacket
    /// </summary>
    public static ZonePowerPacket FromPacket(DataPacket dataPacket)
    {
        if (dataPacket.MessageType != 0x00)
        {
            throw new ArgumentException("Cannot create ZonePowerPacket from packet with MessageType != 0x00");
        }

        var zonePowerPacket = new ZonePowerPacket();
        dataPacket.CopyToPacket(zonePowerPacket);
        return zonePowerPacket;
    }
}